src/main/java/org/spongepowered/api/profile/GameProfileCache.java

```diff
@@ -0,0 +1,313 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.profile;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import org.spongepowered.api.entity.living.player.User;
+import org.spongepowered.api.service.user.UserStorageService;
+import org.spongepowered.api.util.GuavaCollectors;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.UUID;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a cache of {@link GameProfile}s.
+ */
+public interface GameProfileCache {
+
+    /**
+     * Add an entry to this cache.
+     *
+     * @param profile The profile to cache
+     * @return {@code true} if the profile was successfully cached,
+     * otherwise {@code false}
+     */
+    default boolean add(GameProfile profile) {
+        return this.add(profile, null);
+    }
+
+    /**
+     * Add an entry to this cache, with an optional expiration date.
+     *
+     * @param profile The profile to cache
+     * @param expiry The expiration date
+     * @return {@code true} if the profile was successfully cached,
+     * otherwise {@code false}
+     */
+    default boolean add(GameProfile profile, @Nullable Date expiry) {
+        return this.add(profile, false, expiry);
+    }
+
+    /**
+     * Add an entry to this cache, with an optional expiration date.
+     *
+     * @param profile The profile to cache
+     * @param overwrite If we should overwrite the cache entry for
+     * the provided profile
+     * @param expiry The expiration date
+     * @return {@code true} if the profile was successfully cached,
+     * otherwise {@code false}
+     */
+    boolean add(GameProfile profile, boolean overwrite, @Nullable Date expiry);
+
+    /**
+     * Gets a {@link GameProfile} from this cache by its unique id.
+     *
+     * @param uniqueId The unique id of the profile
+     * @return The profile, if present, or {@link Optional#empty()} if
+     * the cache did not contain a profile with the provided id
+     */
+    default Optional<GameProfile> get(UUID uniqueId) {
+        checkNotNull(uniqueId, "unique id");
+        Collection<GameProfile> profiles = this.getBulkId(Collections.singleton(uniqueId));
+        if (profiles.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.ofNullable(Iterables.getFirst(profiles, null));
+    }
+
+    /**
+     * Gets {@link GameProfile}s in bulk by their unique id.
+     *
+     * @param uniqueIds The unique ids
+     * @return A collection of successfully found up profiles
+     */
+    Collection<GameProfile> getBulkId(Iterable<UUID> uniqueIds);
```

Comment:
Maybe just `Map<UUID,GameProfile>`. You could skip uuids that aren't present. That way you can check is the key present and don't have to check on the result for null and absent.
But this is just a matter of preference.